codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ValidationGenerator.java

```diff
@@ -227,13 +233,49 @@ private void generateShapeValidationFunctions(
                             writer.openBlock("if err := $L($L); err != nil {", "}", helperName, valueVar, () -> {
                                 writer.addUseImports(SmithyGoDependency.SMITHY);
                                 writer.write("invalidParams.AddNested(fmt.Sprintf(\"[%q]\", key), "
-                                                + "err.(smithy.InvalidParamsError))");
+                                        + "err.(smithy.InvalidParamsError))");
                             });
                         });
                         break;
 
                     case UNION:
-                        // TODO: Implement Union support
+                        UnionShape unionShape = shape.asUnionShape().get();
+                        Symbol unionSymbol = symbolProvider.toSymbol(unionShape);
+
+                        Set<MemberShape> memberShapes = unionShape.getAllMembers().values().stream()
+                                .filter(memberShape ->
+                                        shapesWithHelpers.contains(model.expectShape(memberShape.getTarget()).getId()))
+                                .collect(Collectors.toCollection(TreeSet::new));
+
```

Comment:
Will this filter the member shapes to only those will with validation traits?

Reply:
Correct, this filters to only include members who refer to a shape that either have a member marked as required, or are an intermediate shape that can contain references to another shape that has such a trait.